### ISSUE TITLE
Introduce Haskell infra and Dispute onchain script

### ISSUE THREAD

**Issue by uhbif19** [eLearningDAO/POCRE#365]
Part of #362

Domain design of script:

* Fixed part of Dispute consist of dispute claim (material and claimed author), list of jury, fixed Hydra Head and vote interval (accepted timing).
https://github.com/eLearningDAO/POCRE/pull/366/files#diff-1fce2ef361d23beb37c1c6562855513122a40a8c236c2e52cbeb16eb903e9744R45
* There are two possible actions on Dispute: Vote and Settle.
   Vote can be done only once by signer in jury list and in vote interval.
   `Settle AllVotesCasted` could only be done when all votes are casted.
   `Settle Timeout` could only be done after vote interval, all not-voted jury will be considered `Abstain`
https://github.com/eLearningDAO/POCRE/pull/366/files#diff-1fce2ef361d23beb37c1c6562855513122a40a8c236c2e52cbeb16eb903e9744R34
* There are three possible `Decision`s which could be casted: Yes, No and Abstain.
   Result is always Yes or No. It is Yes iff `votes for Yes > votest for No`.
* All votes casted are recorded in UTxO datum and persist on-chain after settlement.

Technical design:

* Default StateToken security scheme are used, so API/frontend should check for token in `Dispute.tokenAssetClass` present for check of UTxO correctness.
* Result is encoded in UTxO with `DisputeDatum` which `state` is `Settled SomeReason`.
* Jury and votes list are recorded with `PubKeyHash`.

**Comment by AltiMario** [comment]
About Abstain would be nice to have @huzaifa-99 on that if he is around for a more precise reply.
What is the most basic option for the quorum required?
"authorship change" should be addressed.

**Comment by GeorgeFlerovsky** [comment]
@AltiMario How is authorship managed in the existing POCRE implementation?

**Comment by uhbif19** [comment]
@GeorgeFlerovsky It is managed through metadata but I cannot get details, cuz seems that all metadata descriptions are in imperative code and I cannot find schema for them, so it will take time to get it.

**Comment by uhbif19** [comment]
@huzaifa-99 

So my current plan for interaction with frontend/backend part is as follows:

1. Hydra cluster is created statically and used only once, cuz now Hydra cannot not handle reuse/recreation
2. Frontends do get WS address for (random?) Delegate server on Hydra cluster and connects to it
3. Frontend initializes Dispute with terms (timing, material id, claimed author) for it by creating L1 voting script UTxO.
    Probably that should be claimed author who do that?
    Then it move vote UTxO to L2 by calling Delegate over WS connection.
4. Jury cast votes on L2 by calling Delegate over WS connection.
5. Once voting could be settled (timeout or all votes cast), delegate server (any) creates Settle transaction,
    closes Hydra Head, and UTxO returns back to L1 in settled state.
6. Then delegate server (any) adds metadata to this L1 UTxO, so 

Comments and problems:

1. As for init UTxO I think there is no way to check that such request has correct terms (jury list and such)
    in current arch.
    One could just have fixed server-side signature which identifies that this dispute is correct.
    But given that jury are random and taken from DB there is no simple way to do.
    Or jury can just vote "Abstain" or I can add another "Invalid" Decision for jury,
    so they cast it if the terms are created wrong.
2. The main problem I see is serialization on client. I said that it could be JSON, but I was wrong.
    Sent commands should be signed L1 datum, so client should have a way to serialize them.
    I am discussing how to do that simplest possible way.
3. Last question is how to deploy Hydra cluster and publish Delegate server ports.
    We use docker-compose, but I am not sure how that fits your arch.

**Comment by AltiMario** [comment]
about comments and problems:
1- We could simplify having fixed jury members and always the same. 
2- ok keep us informed
3- we deployed on netlify and the db is on AWS

**Comment by uhbif19** [comment]
>  Hope all these things clarify the current approach and areas to improve.

Thank you for detailed answer.

Scope of my task is only voting on litigation, so I hope that most of this flow design you described does not affect
the logic I am trying to do. AFAIU `START_LITIGATION` should be left on same backend code.

So far I have two domain specific questions:

1. Looks like both creation and material can be disputed/voted for authorship, and voting works the same. Is that right?
2. Which metadata should be created as result of voting? It looks like `ACCEPT_RECOGNITION` is it, but I am not sure.
https://github.com/eLearningDAO/POCRE/blob/e6d540fe60761acf8cb23aa23acd937af1d2b807/app/web-frontend/src/pages/Recognitions/common/hooks/useRecognitions.jsx#L136

> I don't know much about Hydra and how it works with web3 in general

Yes, but I propose interface which should be mostly independent form Hydra specifics.
I found solution for serialization, so I will compile small JS lib for serializing datums, which will not depend on Hydra specifics.
So from frontend/backend side question is how to provide delegate WS address and talk to it.
As for delegate communication frontend should just call mine datum serialization function, probably passing it Cip30 instance, and then send it over WS connection.
For other details I will try to show usage snippets and/or JSON schemas.

**Comment by uhbif19** [comment]
@huzaifa-99 

> It should be transactionPurposes.CAST_LITIGATION_VOTE, 

But this is not vote. All voting will happen on Hydra. This is final decision.

**Comment by uhbif19** [comment]
@huzaifa-99 

So if my code would create `ACCEPT_RECOGNITION` tx in case of vote success and no Tx in case of vote unsuccessful, would it be correct behavior for your backend to work with it?
